File: backend/src/main/java/com/company/pmo/module/notification/Notification.java
package com.company.pmo.module.notification;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;

/**
 * 通知中心持久化(P1.5 收尾)
 *
 * 解决原来 NotificationListener 只发邮件(用户不在邮件中也能看到)
 * - 事件触发时:写一行(status=UNREAD),再发邮件
 * - 前端右上角铃铛:轮询 /unread-count
 * - 点击铃铛:list 分页
 * - 点单条:mark-read
 * - 批量:mark-all-read
 */
@Entity
@Table(name = "notification", indexes = {
        @Index(name = "ix_notification_recipient_unread", columnList = "recipient_id,status,created_at"),
        @Index(name = "ix_notification_recipient_created", columnList = "recipient_id,created_at")
})
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class Notification {

    public enum NotificationStatus { UNREAD, READ }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** 谁接收(永远 = 当前操作者的 user.id) */
    @Column(name = "recipient_id", nullable = false)
    private Long recipientId;

    /** 业务分类:INITIATION_SUBMIT / INITIATION_DECIDE / INITIATION_SUPPLEMENT */
    @Column(nullable = false, length = 32)
    private String category;

    /** 关联业务资源(立项 id / 审批 id) */
    @Column(name = "resource_id")
    private Long resourceId;

    /** 业务编号(冗余,方便列表展示不用 JOIN) */
    @Column(name = "resource_code", length = 64)
    private String resourceCode;

    @Column(nullable = false, length = 256)
    private String title;

    @Column(columnDefinition = "text")
    private String content;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 16)
    private NotificationStatus status;

    @Column(name = "read_at")
    private Instant readAt;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;
}

File: frontend/src/api/notification.ts
/**
 * 通知中心 API (P1.5 收尾)
 *   GET  /api/notifications/unread-count
 *   GET  /api/notifications/page?status=&page=&size=
 *   POST /api/notifications/read   { all?: true } | { ids: number[] }
 */
import api from './client'

export type NotificationStatus = 'UNREAD' | 'READ'
export type NotificationCategory =
  | 'INITIATION_SUBMIT'
  | 'INITIATION_DECIDE'
  | 'INITIATION_SUPPLEMENT'
  | 'TIMESHEET_SUBMIT'
  | 'TIMESHEET_DECIDE'
  | 'TIMESHEET_BATCH_APPROVED'
  | 'TIMESHEET_REMINDER'

export interface Notification {
  id: number
  recipientId: number
  category: NotificationCategory
  resourceId: number
  resourceCode: string
  title: string
  content: string
  status: NotificationStatus
  readAt: string | null
  createdAt: string
}

export interface NotificationPage {
  rows: Notification[]
  total: number
  page: number
  size: number
}

export async function fetchUnreadCount(): Promise<{ count: number }> {
  return api.get('/notifications/unread-count')
}

export async function fetchNotificationPage(
  status: NotificationStatus | 'ALL' = 'ALL',
  page = 0,
  size = 20
): Promise<NotificationPage> {
  const params: Record<string, string | number> = { page, size }
  if (status !== 'ALL') params.status = status
  return api.get('/notifications/page', { params })
}

export async function markRead(body: { all?: true } | { ids: number[] }): Promise<number> {
  return api.post('/notifications/read', body)
}

/** 分类中文名(简单映射,够用即可) */
export const CATEGORY_LABEL: Record<NotificationCategory, string> = {
  INITIATION_SUBMIT: '立项提交',
  INITIATION_DECIDE: '审批决定',
  INITIATION_SUPPLEMENT: '补料重提',
  TIMESHEET_SUBMIT: '工时提交',
  TIMESHEET_DECIDE: '工时审批',
  TIMESHEET_BATCH_APPROVED: '工时批量审批',
  TIMESHEET_REMINDER: '工时催办',
}
